File: packages/igloo-ts/src/message.ts
import type { MessageId, Message } from 'iced:app/message@0.1.0';

/**
 * Manages message registration and dispatch for the Elm-architecture pattern.
 *
 * The MessageManager maps MessageId values (bigint) to handler functions that
 * produce application messages when events occur (e.g., button clicks).
 *
 * @typeParam Msg - The application's message type
 */
export class MessageManager<Msg> {
  private nextId: bigint = 0n;
  private handlers: Map<bigint, (message: Message) => Msg> = new Map();

  /**
   * Register a handler that produces a message when triggered.
   * Returns a MessageId that can be passed to widgets.
   */
  register(handler: (message: Message) => Msg): MessageId {
    const id = this.nextId++;
    this.handlers.set(id, handler);
    return id;
  }

  /**
   * Register a simple handler that ignores the Message payload.
   * Useful for simple button clicks that don't need event data.
   */
  on(handler: () => Msg): MessageId {
    return this.register(() => handler());
  }

  /**
   * Dispatch a message by its ID, returning the application message if found.
   */
  dispatch(id: MessageId, message: Message): Msg | undefined {
    const handler = this.handlers.get(id);
    return handler?.(message);
  }

  /**
   * Clear all registered handlers. Called before each view() to avoid stale handlers.
   */
  clear(): void {
    this.handlers.clear();
    this.nextId = 0n;
  }
}

File: packages/igloo-py/src/igloo_py/message.py
"""Message manager for the Elm-architecture pattern."""

from typing import TypeVar, Generic, Callable, Optional

# These types will be available after generating bindings
# For now, we use simple type aliases
MessageId = int
Message = dict

Msg = TypeVar("Msg")


class MessageManager(Generic[Msg]):
    """
    Manages message registration and dispatch for the Elm-architecture pattern.

    The MessageManager maps MessageId values to handler functions that
    produce application messages when events occur (e.g., button clicks).
    """

    def __init__(self) -> None:
        self._next_id: int = 0
        self._handlers: dict[int, Callable[[Message], Msg]] = {}

    def register(self, handler: Callable[[Message], Msg]) -> MessageId:
        """
        Register a handler that produces a message when triggered.
        Returns a MessageId that can be passed to widgets.
        """
        msg_id = self._next_id
        self._next_id += 1
        self._handlers[msg_id] = handler
        return msg_id

    def on(self, handler: Callable[[], Msg]) -> MessageId:
        """
        Register a simple handler that ignores the Message payload.
        Useful for simple button clicks that don't need event data.
        """
        return self.register(lambda _: handler())

    def dispatch(self, msg_id: MessageId, message: Message) -> Optional[Msg]:
        """
        Dispatch a message by its ID, returning the application message if found.
        """
        handler = self._handlers.get(msg_id)
        return handler(message) if handler else None

    def clear(self) -> None:
        """
        Clear all registered handlers.
        Called before each view() to avoid stale handlers.
        """
        self._handlers.clear()
        self._next_id = 0
